core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/snapshot/DefaultProgressSnapshot.java

```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Override
+    public Optional<Instant> startTime() {
+        return this.startTime;
+    }
+
+    @Override
+    public Optional<Duration> elapsedTime() {
+        return this.startTime.isPresent() ? Optional.of(Duration.between(startTime.get(), Instant.now())) : Optional.empty();
+    }
+
+    @Override
+    public Optional<Duration> estimatedTimeRemaining() {
+        if (!elapsedTime().isPresent() || !remainingBytes().isPresent()) {
+            return Optional.empty();
+        }
+
+        long remainingTime = remainingBytes().getAsLong() * elapsedTime().get().toMillis() / transferredBytes;
+        return Optional.of(Duration.ofMillis(remainingTime));
+
+    }
+
+    @Override
+    public OptionalDouble averageBytesPer(TimeUnit timeUnit) {
+        if (!this.elapsedTime().isPresent()) {
+            return OptionalDouble.empty();
+        }
+
+        return this.elapsedTime().get().equals(Duration.ZERO) ? OptionalDouble.of(1.0) :
+               OptionalDouble.of((double) this.transferredBytes / timeUnit.convert(elapsedTime().get().toMillis(), timeUnit));
+    }
+
+    @Override
+    public OptionalLong totalBytes() {
+        return totalBytes == null ? OptionalLong.empty() : OptionalLong.of(totalBytes);
+    }
+
+    @Override
+    public OptionalDouble ratioTransferred() {
+        if (totalBytes == null) {
+            return OptionalDouble.empty();
+        }
+        return totalBytes == 0 ? OptionalDouble.of(1.0) : OptionalDouble.of(transferredBytes / totalBytes.doubleValue());
+    }
+
+    @Override
+    public OptionalLong remainingBytes() {
+        if (totalBytes == null) {
+            return OptionalLong.empty();
+        }
+        return totalBytes == 0 ? OptionalLong.of(0) : OptionalLong.of(totalBytes - transferredBytes);
+    }
+
+    @Override
+    public DefaultProgressSnapshot.Builder toBuilder() {
+        return new Builder(this);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder implements CopyableBuilder<Builder, DefaultProgressSnapshot> {
+        private long transferredBytes;
+        private Long totalBytes;
+        private Optional<Instant> startTime = Optional.empty();
+
+        private Builder() {
+        }
+
+        private Builder(DefaultProgressSnapshot progressSnapshot) {
+            this.transferredBytes = progressSnapshot.transferredBytes;
+            this.totalBytes = progressSnapshot.totalBytes;
+            this.startTime = progressSnapshot.startTime;
+        }
+
+        public Builder transferredBytes(Long transferredBytes) {
+            this.transferredBytes = transferredBytes;
+            return this;
+        }
+
+        public Long getTransferredBytes() {
+            return this.transferredBytes;
+        }
+
+        public Builder totalBytes(Long totalBytes) {
+            this.totalBytes = totalBytes;
+            return this;
+        }
+
+        public Long getTotalBytes() {
```

Comment:
Can we remove getter methods? We can get value from member variables. It's weird that we have the same in `DefaultTransferProgressSnapshot` 😓

Reply:
yeah I can remove them. I did wonder why we needed it there, but i thought it could be useful for testing. But yes, dont think it makes sense to have them.